lib/ddtrace/contrib/rack/middleware_tracing.rb

```diff
@@ -0,0 +1,90 @@
+module Datadog
+  module Contrib
+    module Rack
+      # A module that injects tracing into the middleware stack itself.
+      #
+      # It's rather complicated due to the limited introspective capabilities of Rack
+      # middlewares, as well as the requirement that middleware spans not be nested.
+      # That is, instead of this:
+      #
+      #   [AAAAAAAAAAAAAAAAAAAAA]
+      #     [BBBBBBBBBBBBBBBBBB]
+      #        [CCCCCCCC]
+      #
+      # We want this:
+      #
+      #   [A][BB][CCCCCCC][BBBB][A]
```

Comment:
I'm not sure if this is something that should be implemented at this layer. I completely understand however the rationale behind it. 
I think it should be possible to handle it on UI layer though.

Reply:
I'd much prefer this to be handled at the UI layer as well. One issue though: in my local version of the code, the middleware spans have a separate service, which enables us to get an APM page for each middleware. Would that work with nested spans?

Reply:
I think it should. You could assign each middleware span whatever service name you want if you want each to appear as a different service.

Reply:
Nah, I want a single service called `my-app-middleware` with each middleware represented as a resource.